{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib notebook\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import string\n",
    "import seaborn as sns\n",
    "\n",
    "import scipy as sp\n",
    "from scipy import sparse\n",
    "import sklearn\n",
    "\n",
    "## add your packages ##\n",
    "\n",
    "import time\n",
    "import pickle\n",
    "from openTSNE import TSNE\n",
    "\n",
    "# import memory_profiler\n",
    "\n",
    "# %load_ext memory_profiler\n",
    "\n",
    "from pathlib import Path\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import black\n",
    "import jupyter_black\n",
    "\n",
    "jupyter_black.load(line_length=79)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variables_path = Path(\"../results/variables\")\n",
    "figures_path = Path(\"../results/figures\")\n",
    "data_path = Path(\"../data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use(\"matplotlib_style.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(42)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data\n",
    "\n",
    "We are going to use the multimodal data from the paper Scala et al. 2021 (https://www.nature.com/articles/s41586-020-2907-3#Sec7). In particular, you will work with transcriptomics and electrophysiological data. From the transcriptomics gene counts, we will only work with the exon counts for simplicity. Some of the electrophysiological features are not high-quality recordings, therefore we will also filter them out for the project."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# META DATA\n",
    "\n",
    "meta = pd.read_csv(data_path / \"m1_patchseq_meta_data.csv\", sep=\"\\t\")\n",
    "\n",
    "cells = meta[\"Cell\"].values\n",
    "\n",
    "layers = meta[\"Targeted layer\"].values.astype(\"str\")\n",
    "cre = meta[\"Cre\"].values\n",
    "yields = meta[\"Yield (pg/µl)\"].values\n",
    "yields[yields == \"?\"] = np.nan\n",
    "yields = yields.astype(\"float\")\n",
    "depth = meta[\"Soma depth (µm)\"].values\n",
    "depth[depth == \"Slice Lost\"] = np.nan\n",
    "depth = depth.astype(float)\n",
    "thickness = meta[\"Cortical thickness (µm)\"].values\n",
    "thickness[thickness == 0] = np.nan\n",
    "thickness = thickness.astype(float)\n",
    "traced = meta[\"Traced\"].values == \"y\"\n",
    "exclude = meta[\"Exclusion reasons\"].values.astype(str)\n",
    "exclude[exclude == \"nan\"] = \"\"\n",
    "\n",
    "mice_names = meta[\"Mouse\"].values\n",
    "mice_ages = meta[\"Mouse age\"].values\n",
    "mice_cres = np.array(\n",
    "    [\n",
    "        c if c[-1] != \"+\" and c[-1] != \"-\" else c[:-1]\n",
    "        for c in meta[\"Cre\"].values\n",
    "    ]\n",
    ")\n",
    "mice_ages = dict(zip(mice_names, mice_ages))\n",
    "mice_cres = dict(zip(mice_names, mice_cres))\n",
    "\n",
    "print(\"Number of cells with measured depth:    \", np.sum(~np.isnan(depth)))\n",
    "print(\"Number of cells with measured thickness:\", np.sum(~np.isnan(thickness)))\n",
    "print(\"Number of reconstructed cells:          \", np.sum(traced))\n",
    "\n",
    "sliceids = meta[\"Slice\"].values\n",
    "a, b = np.unique(sliceids, return_counts=True)\n",
    "assert np.all(b <= 2)\n",
    "print(\"Number of slices with two cells:        \", np.sum(b == 2))\n",
    "\n",
    "# Some consistency checks\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Date\"].values[mice_names == m]).size == 1\n",
    "        for m in mice_names\n",
    "    ]\n",
    ")\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Mouse age\"].values[mice_names == m]).size == 1\n",
    "        for m in mice_names\n",
    "    ]\n",
    ")\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Mouse gender\"].values[mice_names == m]).size == 1\n",
    "        for m in mice_names\n",
    "    ]\n",
    ")\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Mouse genotype\"].values[mice_names == m]).size == 1\n",
    "        for m in mice_names\n",
    "    ]\n",
    ")\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Mouse\"].values[sliceids == s]).size == 1\n",
    "        for s in sliceids\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transcriptomic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# READ COUNTS\n",
    "\n",
    "data_exons = pd.read_csv(\n",
    "    data_path / \"m1_patchseq_exon_counts.csv.gz\", na_filter=False, index_col=0\n",
    ")\n",
    "exonCounts = data_exons.values.transpose()\n",
    "\n",
    "assert all(cells == data_exons.columns)\n",
    "genes = np.array(data_exons.index)\n",
    "\n",
    "print(\"Count matrix shape (exon):  \", exonCounts.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENE LENGTH\n",
    "\n",
    "data = pd.read_csv(data_path / \"gene_lengths.txt\")\n",
    "assert all(data[\"GeneID\"] == genes)\n",
    "exonLengths = data[\"exon_bp\"].values\n",
    "intronLengths = data[\"intron_bp\"].values\n",
    "print(exonLengths)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cluster colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_colors = np.load(data_path / \"cluster_colors.npy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_colors.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_type = np.load(data_path / \"rna_type.npy\", allow_pickle=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_type.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle_in = open(data_path / \"dict_rna_type_colors.pkl\", \"rb\")\n",
    "dict_rna_type_colors = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(dict_rna_type_colors)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Electrophysiological features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EPHYS DATA\n",
    "\n",
    "ephysData = pd.read_csv(data_path / \"m1_patchseq_ephys_features.csv\")\n",
    "ephysNames = np.array(ephysData.columns[1:]).astype(str)\n",
    "ephysCells = ephysData[\"cell id\"].values\n",
    "ephysData = ephysData.values[:, 1:].astype(\"float\")\n",
    "names2ephys = dict(zip(ephysCells, ephysData))\n",
    "ephysData = np.array(\n",
    "    [\n",
    "        names2ephys[c] if c in names2ephys else ephysData[0] * np.nan\n",
    "        for c in cells\n",
    "    ]\n",
    ")\n",
    "\n",
    "print(\"Number of cells with ephys data:\", np.sum(np.isin(cells, ephysCells)))\n",
    "\n",
    "assert np.sum(~np.isin(ephysCells, cells)) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering ephys data\n",
    "\n",
    "features_exclude = [\n",
    "    \"Afterdepolarization (mV)\",\n",
    "    \"AP Fano factor\",\n",
    "    \"ISI Fano factor\",\n",
    "    \"Latency @ +20pA current (ms)\",\n",
    "    \"Wildness\",\n",
    "    \"Spike frequency adaptation\",\n",
    "    \"Sag area (mV*s)\",\n",
    "    \"Sag time (s)\",\n",
    "    \"Burstiness\",\n",
    "    \"AP amplitude average adaptation index\",\n",
    "    \"ISI average adaptation index\",\n",
    "    \"Rebound number of APs\",\n",
    "]\n",
    "features_log = [\n",
    "    \"AP coefficient of variation\",\n",
    "    \"ISI coefficient of variation\",\n",
    "    \"ISI adaptation index\",\n",
    "    \"Latency (ms)\",\n",
    "]\n",
    "\n",
    "X = ephysData\n",
    "print(X.shape)\n",
    "for e in features_log:\n",
    "    X[:, ephysNames == e] = np.log(X[:, ephysNames == e])\n",
    "X = X[:, ~np.isin(ephysNames, features_exclude)]\n",
    "\n",
    "keepcells = ~np.isnan(np.sum(X, axis=1))\n",
    "X = X[keepcells, :]\n",
    "print(X.shape)\n",
    "\n",
    "X = X - X.mean(axis=0)\n",
    "ephysData_filtered = X / X.std(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.sum(np.isnan(ephysData_filtered))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Research questions to investigate\n",
    "\n",
    "**1) Inspect the data computing different statistics.** Keep in mind that the data is read counts, not UMI, so it is not supposed to follow a Poisson distribution.\n",
    "\n",
    "**2) Normalize and transform the data.** There are several ways of normalizing the data (Raw, CPM, CPMedian, RPKM, see https://www.reneshbedre.com/blog/expression_units.html, https://translational-medicine.biomedcentral.com/articles/10.1186/s12967-021-02936-w). Take into account that there are certain normalizations that only make sense for UMI data. You also explored different transformations in the assignment (none, log, sqrt). Compare how the different transformations change the two-dimensional visualization.\n",
    "\n",
    "**3) Two-dimensional visualization.** Try different methods (t-SNE, UMAP) / parameters (exagg., perplex.) for visualizations. Compare them using quantitative metrics (e.g., distance correlation, kNN accuracy/recall in high-dim vs. two-dim). Think about also using the electrophysiological features for different visualizations.\n",
    "\n",
    "**4) Clustering.** Try different clustering methods (leiden, GMM). Implement a negative binomial mixture model. For that you can follow a similar method that what is described in Harris et al. 2018 (https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2006387#abstract0), with fixed r (r=2) and S (set of important genes). Evaluate your clustering results (metrics, compare number of clusters to original labels,...).\n",
    "\n",
    "**5) Correlation in between electrophysiological features and genes/PCs.** Find correlations and a way of visualizing them.\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1) Inspecting Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for nan\n",
    "assert np.isnan(exonCounts).sum() == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kick out all genes with all counts = 0\n",
    "genes = genes[exonCounts.sum(axis=0) > 0]\n",
    "exonLengths = exonLengths[exonCounts.sum(axis=0) > 0]\n",
    "exonCounts = exonCounts[:, exonCounts.sum(axis=0) > 0]\n",
    "\n",
    "# check if they keep same size\n",
    "assert genes.size == exonCounts.shape[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the average expression for each gene\n",
    "avg_counts = np.mean(exonCounts, axis=0)\n",
    "\n",
    "# Compute the fraction of zeros for each gene\n",
    "zero_fraction = np.sum(exonCounts == 0, axis=0) / exonCounts.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the Poisson prediction\n",
    "\n",
    "Poiss_predicted_fraction = np.exp(-np.unique(avg_counts))\n",
    "\n",
    "\n",
    "# Plot\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 4))\n",
    "ax.scatter(avg_counts, zero_fraction, s=5, label=\"Data\")\n",
    "ax.plot(\n",
    "    np.unique(avg_counts),\n",
    "    Poiss_predicted_fraction,\n",
    "    c=\"red\",\n",
    "    label=\"Possion Prediction\",\n",
    ")\n",
    "\n",
    "ax.set_ylabel(\"Fraction of zero expression\")\n",
    "ax.set_xlabel(\"Mean read counts\")\n",
    "ax.set_title(\"Fraction of zeros for each gene vs Poisson prediction\")\n",
    "\n",
    "ax.legend()\n",
    "ax.set_xscale(\"log\")\n",
    "# add plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chen, W., Li, Y., Easton, J. et al. UMI-count modeling and differential expression analysis for single-cell RNA sequencing. Genome Biol 19, 70 (2018). https://doi.org/10.1186/s13059-018-1438-9\n",
    "\n",
    "Jiang, R., Sun, T., Song, D. et al. Statistics or biology: the zero-inflation controversy about scRNA-seq data. Genome Biol 23, 31 (2022). https://doi.org/10.1186/s13059-022-02601-\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we were expecting, the read counts seem to deviate significantly from the Poisson model. First observation is that the fraction of zeroes observed is consistently higher than a poisson model would predict. Another one is a visually noticeable variability of the mean read counts for a same fraction of zeros, resulting in a wide curve (very different from what was observed in the dataset of CodingLab 7). One of the main factors underlying this deviances is probably the so-called amplification bias. Since the amount of transcripts collected from single cells is very small, cDNA amplification is necessary for the construction of the library. Nevertheless, this amplification is also a stochastic process, resulting in distortions (e.g., the same amount of a specific transcript was captured for two different cells. But in one of them, the cDNA amplification lead to a higher amount of read counts) (Chen et al. 2018). The UMI techniques try to mitigate this bias, by tagging the transcripts befor amplification.\n",
    "\n",
    "Another point raised by Chen et al. (2018) is the higher number of dropout events in single cell sequencing (missing values, due to biological and technical stochasticities). The authors argue that accounting for these kind of event is relevant for read counts, but not much for UMI. Apparently, there is a big discussion in the scRNA-seq field if models should include zero inflations (to correct for the dropout events) or not (Jiang et al. 2022). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Some statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "depth = np.sum(exonCounts, axis=1)\n",
    "genes_detected = np.sum(exonCounts != 0, axis=1)\n",
    "\n",
    "# Basic statistics (same reported in GitHub https://github.com/berenslab/mini-atlas/blob/master/code/ttype-assignment.ipynb)\n",
    "print(\n",
    "    \"Depth: median {:.0f}, mean+-STD log10: {:.1f}+-{:.1f}\".format(\n",
    "        np.median(depth), np.mean(np.log10(depth)), np.std(np.log10(depth))\n",
    "    )\n",
    ")\n",
    "\n",
    "\n",
    "print(\n",
    "    \"Genes detected per cell: median {:.0f}, mean+-STD: {:.0f}+-{:.0f}\".format(\n",
    "        np.mean(genes_detected),\n",
    "        np.mean(genes_detected),\n",
    "        np.std(genes_detected),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "# Plot number of detected genes vs number of counts\n",
    "axs[0].scatter(depth, genes_detected, s=10)\n",
    "axs[0].set_xscale(\"log\")\n",
    "axs[0].set_ylim(0, 20000)\n",
    "axs[0].set_ylabel(\"Number of detected Genes\")\n",
    "axs[0].set_xlabel(\"Sum of counts\")\n",
    "axs[0].set_title(\"\")\n",
    "\n",
    "# Plot histogram of sequencing depths (1 pt)\n",
    "axs[1].hist(depth, bins=20)\n",
    "axs[1].set_xlabel(\"Sum of counts per cell\")\n",
    "axs[1].set_ylabel(\"Number of Cells\")\n",
    "axs[1].set_title(\"Sequencing depth per cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Mean-Variance analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the variance of the expression counts of each gene\n",
    "var_counts = np.var(exonCounts, axis=0)\n",
    "\n",
    "# Fano factor\n",
    "Fano = var_counts / avg_counts\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "axs[0].scatter(avg_counts, var_counts, s=10)\n",
    "axs[0].plot(\n",
    "    np.arange(0, np.max(avg_counts), 1),\n",
    "    np.arange(0, np.max(avg_counts), 1),\n",
    "    color=\"r\",\n",
    "    linewidth=2,\n",
    "    label=\"y = x\",\n",
    ")\n",
    "\n",
    "axs[0].set_xscale(\"log\")\n",
    "axs[0].set_yscale(\"log\")\n",
    "axs[0].set_xlabel(\"Mean Expression\")\n",
    "axs[0].set_ylabel(\"Expression Variance\")\n",
    "axs[0].set_title(\"Gene variance vs mean expression\")\n",
    "axs[0].tick_params(which=\"minor\", length=1, color=\"k\")\n",
    "axs[0].tick_params(which=\"major\", length=5, color=\"k\")\n",
    "axs[0].legend()\n",
    "\n",
    "# Fano Plot\n",
    "axs[1].scatter(avg_counts, Fano, s=10)\n",
    "axs[1].plot(\n",
    "    np.arange(0, np.max(avg_counts), 1),\n",
    "    np.ones(np.arange(0, np.max(avg_counts), 1).size),\n",
    "    color=\"r\",\n",
    "    linewidth=2,\n",
    "    label=\"Model: F = 1\",\n",
    ")\n",
    "axs[1].set_xlabel(\"Mean Expression\")\n",
    "axs[1].set_ylabel(\"Fano Factor\")\n",
    "axs[1].set_title(\"Fano Factor vs mean expression\")\n",
    "# axs[1].set_ylim(0.4, 200)\n",
    "axs[1].set_xscale(\"log\")\n",
    "axs[1].set_yscale(\"log\")\n",
    "axs[1].tick_params(which=\"minor\", length=2, color=\"k\")\n",
    "axs[1].tick_params(which=\"major\", length=5, color=\"k\")\n",
    "\n",
    "axs[1].legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geneSelection(\n",
    "    data,\n",
    "    threshold=0,\n",
    "    atleast=10,\n",
    "    yoffset=0.02,\n",
    "    xoffset=5,\n",
    "    decay=1.5,\n",
    "    n=None,\n",
    "    plot=True,\n",
    "    markers=None,\n",
    "    genes=None,\n",
    "    figsize=(6, 3.5),\n",
    "    markeroffsets=None,\n",
    "    labelsize=10,\n",
    "    alpha=1,\n",
    "):\n",
    "    zeroRate = 1 - np.mean(data > threshold, axis=0)\n",
    "    meanExpr = np.zeros_like(zeroRate) * np.nan\n",
    "    detected = zeroRate < 1\n",
    "    mask = data[:, detected] > threshold\n",
    "    logs = np.zeros_like(data[:, detected]) * np.nan\n",
    "    logs[mask] = np.log2(data[:, detected][mask])\n",
    "    meanExpr[detected] = np.nanmean(logs, axis=0)\n",
    "\n",
    "    lowDetection = (\n",
    "        np.array(np.sum(data > threshold, axis=0)).squeeze() < atleast\n",
    "    )\n",
    "    zeroRate[lowDetection] = np.nan\n",
    "    meanExpr[lowDetection] = np.nan\n",
    "\n",
    "    if n is not None:\n",
    "        up = 10\n",
    "        low = 0\n",
    "        for t in range(100):\n",
    "            nonan = ~np.isnan(zeroRate)\n",
    "            selected = np.zeros_like(zeroRate).astype(bool)\n",
    "            selected[nonan] = (\n",
    "                zeroRate[nonan]\n",
    "                > np.exp(-decay * (meanExpr[nonan] - xoffset)) + yoffset\n",
    "            )\n",
    "            if np.sum(selected) == n:\n",
    "                break\n",
    "            elif np.sum(selected) < n:\n",
    "                up = xoffset\n",
    "                xoffset = (xoffset + low) / 2\n",
    "            else:\n",
    "                low = xoffset\n",
    "                xoffset = (xoffset + up) / 2\n",
    "        print(\"Chosen offset: {:.2f}\".format(xoffset))\n",
    "    else:\n",
    "        nonan = ~np.isnan(zeroRate)\n",
    "        selected = np.zeros_like(zeroRate).astype(bool)\n",
    "        selected[nonan] = (\n",
    "            zeroRate[nonan]\n",
    "            > np.exp(-decay * (meanExpr[nonan] - xoffset)) + yoffset\n",
    "        )\n",
    "\n",
    "    if plot:\n",
    "        if figsize is not None:\n",
    "            plt.figure(figsize=figsize)\n",
    "        plt.ylim([0, 1])\n",
    "        if threshold > 0:\n",
    "            plt.xlim([np.log2(threshold), np.ceil(np.nanmax(meanExpr))])\n",
    "        else:\n",
    "            plt.xlim([0, np.ceil(np.nanmax(meanExpr))])\n",
    "        x = np.arange(plt.xlim()[0], plt.xlim()[1] + 0.1, 0.1)\n",
    "        y = np.exp(-decay * (x - xoffset)) + yoffset\n",
    "        if decay == 1:\n",
    "            plt.text(\n",
    "                0.4,\n",
    "                0.2,\n",
    "                \"{} genes selected\\ny = exp(-x+{:.2f})+{:.2f}\".format(\n",
    "                    np.sum(selected), xoffset, yoffset\n",
    "                ),\n",
    "                color=\"k\",\n",
    "                fontsize=labelsize,\n",
    "                transform=plt.gca().transAxes,\n",
    "            )\n",
    "        else:\n",
    "            plt.text(\n",
    "                0.4,\n",
    "                0.2,\n",
    "                \"{} genes selected\\ny = exp(-{:.1f}*(x-{:.2f}))+{:.2f}\".format(\n",
    "                    np.sum(selected), decay, xoffset, yoffset\n",
    "                ),\n",
    "                color=\"k\",\n",
    "                fontsize=labelsize,\n",
    "                transform=plt.gca().transAxes,\n",
    "            )\n",
    "\n",
    "        plt.plot(x, y, color=sns.color_palette()[1], linewidth=2)\n",
    "        xy = np.concatenate(\n",
    "            (\n",
    "                np.concatenate((x[:, None], y[:, None]), axis=1),\n",
    "                np.array([[plt.xlim()[1], 1]]),\n",
    "            )\n",
    "        )\n",
    "        t = plt.matplotlib.patches.Polygon(\n",
    "            xy, color=sns.color_palette()[1], alpha=0.4\n",
    "        )\n",
    "        plt.gca().add_patch(t)\n",
    "\n",
    "        plt.scatter(meanExpr, zeroRate, s=1, alpha=alpha, rasterized=True)\n",
    "        if threshold == 0:\n",
    "            plt.xlabel(\"Mean log2 nonzero expression\")\n",
    "            plt.ylabel(\"Frequency of zero expression\")\n",
    "        else:\n",
    "            plt.xlabel(\"Mean log2 nonzero expression\")\n",
    "            plt.ylabel(\"Frequency of near-zero expression\")\n",
    "        plt.tight_layout()\n",
    "\n",
    "        if markers is not None and genes is not None:\n",
    "            if markeroffsets is None:\n",
    "                markeroffsets = [(0, 0) for g in markers]\n",
    "            for num, g in enumerate(markers):\n",
    "                i = np.where(genes == g)[0]\n",
    "                plt.scatter(meanExpr[i], zeroRate[i], s=10, color=\"k\")\n",
    "                dx, dy = markeroffsets[num]\n",
    "                plt.text(\n",
    "                    meanExpr[i] + dx + 0.1,\n",
    "                    zeroRate[i] + dy,\n",
    "                    g,\n",
    "                    color=\"k\",\n",
    "                    fontsize=labelsize,\n",
    "                )\n",
    "        plt.show()\n",
    "    return selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genesForRRR = geneSelection(\n",
    "    exonCounts, n=1000, threshold=32, decay=1.5, yoffset=0.02\n",
    ")\n",
    "print(np.sum(genesForRRR))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalization and Transformation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CTF normalization\n",
    "\n",
    "Write about the different methods. In blog they suggest one for single cell sequencing, but didnt find python implementation. Took CTF from another paper, which claimed it is more robust (https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02568-9). Method is implemented in library (https://pypi.org/project/rnanorm/)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rnanorm import CTF  # pip install rnanorm\n",
    "\n",
    "print(pd.DataFrame(exonCounts).shape)\n",
    "CTF_exon_counts = (\n",
    "    CTF()\n",
    "    .set_output(transform=\"pandas\")\n",
    "    .fit_transform(pd.DataFrame(exonCounts))\n",
    ")\n",
    "CTF_exon_counts = CTF_exon_counts.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "log_CTF_exon_counts = np.log2(CTF_exon_counts + 1)\n",
    "axs[0].scatter(\n",
    "    np.mean(log_CTF_exon_counts, axis=0),\n",
    "    np.var(log_CTF_exon_counts, axis=0),\n",
    "    s=10,\n",
    ")\n",
    "axs[0].plot(\n",
    "    np.arange(0, np.max(np.mean(log_CTF_exon_counts, axis=0)), 1),\n",
    "    np.arange(0, np.max(np.mean(log_CTF_exon_counts, axis=0)), 1),\n",
    "    color=\"r\",\n",
    "    linewidth=2,\n",
    "    label=\"y = x\",\n",
    ")\n",
    "\n",
    "axs[0].set_xscale(\"log\")\n",
    "axs[0].set_yscale(\"log\")\n",
    "axs[0].set_xlabel(\"Mean Expression\")\n",
    "axs[0].set_ylabel(\"Expression Variance\")\n",
    "axs[0].set_title(\"Gene variance vs mean expression\")\n",
    "axs[0].tick_params(which=\"minor\", length=1, color=\"k\")\n",
    "axs[0].tick_params(which=\"major\", length=5, color=\"k\")\n",
    "axs[0].legend()\n",
    "\n",
    "# Fano Plot\n",
    "axs[1].scatter(\n",
    "    np.mean(log_CTF_exon_counts, axis=0),\n",
    "    np.var(log_CTF_exon_counts, axis=0) / np.mean(log_CTF_exon_counts, axis=0),\n",
    "    s=10,\n",
    ")\n",
    "axs[1].plot(\n",
    "    np.arange(0, np.max(np.mean(log_CTF_exon_counts, axis=0)), 1),\n",
    "    np.ones(\n",
    "        np.arange(0, np.max(np.mean(log_CTF_exon_counts, axis=0)), 1).size\n",
    "    ),\n",
    "    color=\"r\",\n",
    "    linewidth=2,\n",
    "    label=\"Model: F = 1\",\n",
    ")\n",
    "axs[1].set_xlabel(\"Mean Expression\")\n",
    "axs[1].set_ylabel(\"Fano Factor\")\n",
    "axs[1].set_title(\"Fano Factor vs mean expression\")\n",
    "axs[1].set_ylim(0.01, 20)\n",
    "axs[1].set_xscale(\"log\")\n",
    "axs[1].set_yscale(\"log\")\n",
    "axs[1].tick_params(which=\"minor\", length=2, color=\"k\")\n",
    "axs[1].tick_params(which=\"major\", length=5, color=\"k\")\n",
    "\n",
    "axs[1].legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In GitHub of original paper, it seems that the normalization applied was: counts / (exonLenght / 1000). So we will also use this one for comparison.\n",
    "\n"
   ]
  },
  {
   "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAboAAACYCAYAAACS5CwMAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAEnQAABJ0Ad5mH3gAAB+USURBVHhe7d1/bBRnmifwLyeB7jKZ5UAM0sKGtE0fdDvS7chik8BmkHHOljEOsi00g8xIsZJNI6yEwZzxBrT2H7aU5BwfJsmsI/eGWWY045udRMbiHGzZg+0jmU3CRNyulLgbrtPuOJBRMhxcZjj2Nvmj732r3uqualf/bpvu4vuRSnTVW7+bfp963nrLtSIqgIiIyKH+jfqXiIjIkRjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoiIjI0RjoaMlE/M1wub1qOI5ZNZ0KYQYnYudWDJ0zajoRJSrJQHfz5zswuH0HLi2oCZpxTGxfgcHecTXuAPOn0WyqzJr9EVVQQpoHEQkFxPAiqtSkklHU538XXtDOawBnmtUkIrJVkoFu7Q9/hm1b3sOHXadwU00L9dYjvGUA+7t2qyklbvo4XDX98Pj1ykwOJ3FhWbMiLSPbexolGF7zph37ovM/hJ/MqxmIqGSUaNOlGw/3DmDN1XZceleMLpzCb88D5c8cwVp9hhIXwU9OjaKycxwvVKtJgsv3dOllRSVpBv6+AFpEkLOe/xfxVJkaIaKSUbr36DYdwV/UA+G/O4iJrnbcqj+PusdU2ZITgWivFyemzfdJmhdd7c92xpu94pmRvqxdM1gsg5q/gLE5LxpqXKrEjr6e2Pot98CM/VOjkswQE/Yh2f4b99aqRGWPuX5U2cwjWY5PDJbtZURtP7ZfmVri8z89iWE0otYU5OzYr1/KZ//i8j+/RCSVdGcUd9d5lF/1I3x1O7Y9mU2T5WdoazqLVTZD2wdqlgwM+16De8q4TxJAz4/ilZWspFph3J8S5R4RMLTKzIVyD3A5lFDrGTzlcM2HcBlbUZ40e5AV6W70eOLrn+28gtYsO3wk23+Xb0St0wtUdGBWbSMSGollNDIotAbNZdbsZzks1flH6Io4breYM7nk64/Lbf90xXB+iZyipAMdFq7glvqYnQcweLYJX9sMg4+oWTLQ4o9X/FV1jfoHzQwmR7zoPrRLjYvyQx2onHsb06J+lVfnBlnhGT3mZGVW6c6gbUzL+Bpxpi++fpfvObRgFJNZXPUn3/8MqePJnepQce7plEElmeU5/3ZZWer1G3Ldv5i8zy8RSSUc6EK4JJsst/hQLjum/DSb3paFyeiSmg8jCHEFX2NUkGKo6RdZms7l3goEw+LqXVR4wUa0BCdFJhZBOCgSCrdR5V9BOFklp2V8icrgrlAfl4HM+mY7ET9Gm6a3u6YQ538upI7H6N04KC4klDTrTyuD5Yv6/BKVmNINdO++jA9lk2XvEOqe8QHn6zEhO6ZkpDAZXWriil01W8UHdYVf5hZX76IiFRUeGg6i1mNkYl5oCUV1rahUAxibyqZqm0doTn1cJkYTZyQ0jm70o6qonuXK/fy7avagMm12nGL9GUm/fHGfX6LSUaKBbhwTx/xAfRce3iRGHxtCreyYcuygyPOKQNnjaKgQV+yvJ6mYysrhEf+Ep0Jw17i0Zq3hiSERqIz7crvg6/Tict9uSweEiP+0fg9OC4SjaDVVfLOdbRiu6IDPdB9neEKVy+fBfKP65yxomU9GzWf6fa/speuMYjQbZvmweb7nv+xpHG6W99gWdyDRpFt/Olkvn+v5JSKpJAOd9swcfKg1PTPnfnIAa+DHZFE8MO7CU+fEVXiwLd40ZanQZTPjKHrG3KiWFasMXCOjIlDFO0DoTVdeUdnGlz+Kx9XjBbI5bVAsE1+/1nEhdq9LbP+VDlQa5TUhHPbncA+u+kW9E0WsiS1e8Wv3ttS2Y9s33TMsDKM5Nrt7j4U4/7LHqaXp0C0uJFRZ+vWnk3755Tm/RPeGFVFBfSYqKNlzsCr0HCJ5VNDaOvqA7qlsmgXvLTIoaj04GQiJbJV2r0tyuAimxwJAxR498yIiygEDHS2tWPNqlvfZ5APubvWsYI6PHzhb/LGH1hE1iYhssemSiIgcjRkdERE5GgMdERE5GgMdERE5GgMdERE5GgMdERE5GgMdERE5GgMdERE5GgMdERE5GgMdERE5GgMdERE5GgMdkfZ3I5O8ey4puUyWf7+TiO6KEg90n+LzJ/4c/7S5CzfVlEXm/x6BzXIefQj4P1UFBHyN2fYIepu+xMdqyr0ngp/sbUOw82TsNUA3f74Dg9t34NKCPq4bx8T2FRiMve9Qvhz3CloZ7IiKnsMzOhEID58C/vocvvvJP2uD1/egKiOS77s7ih504KQv/n6EtT/8GbZteQ8fdp2KXUBpL/vdMoD9ppf9unwn0V1hfdM7ERUfh7+94H8gvHkQ//bX/4ANfJ8ZJZo/jeaat9Fg91LXhVP4bz9ox5qXo6jbZPr8mCo3pFoHERWFkg10/+L/Aa78l6Aa24tNn/RirRqLyy/QaW9uNr3rq8UfwAvV8pNs7tqN0JFBwNeGYa3UXK6zLt+IM6EXUaXGYm/frpsUmcGoNq2ycxwjpswi+fZ1lvKKDsxm8d62L9+6jqFffKPG7kPz2fV4SI0ZPh6IYOSiGhG8z7uw7xE1IiQvl02in+PG/nXASzcQ0ErTLW/dB23/PluNrh130PvSHW3augMbcGjfKu2zlNX+uVbj4MAarFejkvYdjO1Jet5CvSswGfKhHH6E3efRZsrm4vT/C2MN1u9OJ+/jif8fWX43RFRgMtCVtF//TfR/lv9N9H+rUY027T/aDAnzpTA/1BR98Ik3ovNq3Go+evoJT/TBzZ7o8Qtqipx/8/PRGX1UGzfKpJljYn7T+vT5TdPCb0SbNjdFT4e14jTbV+s7Nq3GFq8/Y+9/Ee1p/CL6kRo1fPHmtWjPkZvRL9R4otTl/xqdOTIv1jsfffN9fYo2v2k7ctwokz46KeY3rU+f3zTt2s3oYOO16Mw1rTjt/mnrO/lHNbZ4/ZI8Z01Dqc7Y+ej4o4j+7aPbox98qibZ0L4r03cRNx09bv6OieiucOY9usd71T25V/En8GD9r/X7c9+1zfpSmHsb0yl64pkzLJfvObTgCsJqfpdvxJJ9VdU1qk9mIsszrvTLysWeBhAyby/p9mcwOeJF96Fdalys/1AHKtPsb9YidzB3XX22k6bcnGGt37caXnyD36v51+/baMm+Htpxn/pkJrI8IwvbuBLfEcvfuKYV6JJu/zYCF1fie9+/X42L9X9/NdZZ5o8gbDQIJLNwBbfUx7SCYbHGRLvwQijAN6QT3WV8vCAJGahmO4GeGi9cbjHsPW1TkSUyBSp570YuZwyqedKiuTbWlGlUivHAmWL782EExbZiZXKo6cdlVVwIMhAdPAC886zslSmG9lv4UpVJ6crtmQLV9Vt4XS5nDKp50mLnfabm1Pux76w5cKbY/nURUMW2YmVyePYr3FDFmQnhUlc7bm3xoVx2TPmp0duSiEoNA10KMthE5BV5aBzd6EdVqt51WvDxwq3dC5zBCRF40DmulheD3y6jSy319kVGN6XWHRsK2yFCBpMuEVy6zm7A9/AVhgZuqxJdunILLfisxLo/kyO38ZYIPDiwQS0vhuftMrrUUm9fZHQ/VuuODRtRtVEVp/Puy/jw6nZs6x1C3TM+4Hw9Jt5VZXY85czaiIoUA11GXCj3qI9JzL4uMqrm55IEGhH47DK6jCVsv+xxNFSIjO71NN3aY1llvs96rcJ3NqmPttKVAx//SmRUO1cnCTQi8NlldBlL2P7Gb8HrEhndr1IEXnFOqxu8uDx2wSZTH8fEMT9Q34WH5XofG0JtPRA+dlDkeYkimB4LoFK/wkkgO6MkZONEtOxWyBt16nMJkQ+K78WXc2o0Rt6PM/ewzL3XZWKPR2uvRr2nXY95+82DiPTF75lpPfr6Yv0N0d25FT1j7tg6Yr0uTcuYpd6+ZLMPifNoXd9lk6a1x6fRK/KdRbWvzIL0rCexR2Nir8XU5Tbr37kOXe3xe2bWXp9iuwdW4p3frIytI9br0rSMWbr9s92HxHnU+fEk9GbVelue96H2vSG41TTjcYNb9Qm9L1M+XsBel0TFoEQD3d1mPF5grSDJYDxeYO3uX4y0C4pgroEo1aMFRFQs2HRJ97SqvgzuvyZh91dViKj4MNDRPc6Fp84NomXktaz/qLO/b2v88RAiKlpsuiQiIkdjRkdERI7GQEdERI7GQEdERI7GQEdERI7GQEdERI7GQEdERI7GQEdERI7GQEdERI7GQEdERI7GQEdERI7GQEcOIF+H05zl36qkOLvzJ6eleo9hunKi4sFAl4ObP9+Bwe07cGlBTdCMY2L7Cgz2jqtxWh7yVTltCHaejL0Prhi/H/l+vd72W/hSjRePxedPtwu+zitoTRrM0pUTFQ8Guhys/eHPsG3Le/iw6xRuqmmh3nqEtwxgv/mlnLTk7F6VU6jvp3iDU+GketWQy3cS3RWjaE3yCqN05UTFgoEuJ2483DuANVfbceldMbpwCr89D5Q/cwRr9RloOcyfxtE+oPuVxFfl8PvJSNLzZ3DhqVc6UJn0FUbpyomKA1/Tk4dQ7wpMhnwohx9h93m0LVs2Z7zhfBDwtWFYm+ZF99SIan6yKwda/Nm9Ef3jgQhGLqoR3Ifms+vxkBqT2c7QZ6vRteMOel+6o01bd2ADDu1bpX1OVy4lW7+27C9WWrYH3MZbTTfwe9M6Iv5mVI3tSfp28Fy/H33736gxs5X43o83omqj/Ky/Rf2diFYgWM+PZD0+wbUaBwfWYL0atZRbyow3tK8DXrqBgDbNvG1d4vq9z1vf6J58/bp050+X7i3q6cqJioAMdJSr89HxRxH920e3Rz/4VE3KyEL0UONIdKXNcOh9NUtK89HTT3iiD25uip4O61NmjonxJ94QJZJR7okev6BNiM4PNYnx56Mz+mhaX7x5LfqmaV8+Ojkf7TlyM/qFGpflPY2madduRgcbr0VnrmnFGZUnX/8fo2+KZc3l0fe/sGxfksfcNKQfsb1cvx+ddgwJ29T9a3TmiNjfk39U48bxfhH9SI1rx5NYblpXYrn1+NX6TefL9vzb7psu9fp16c+fTvu/c2xajS2WrpzobmPTZT4WruCW+pidBzB4tglf2wyDpivydFr8RgYHVNU16h9MzBmcy/ccWnAF4QybmNbv22jJDh7acZ/6ZCayGCNL2LgS38E3uHFNK1CSl6de//3w7gQC/3hbjYsM55d3sO4vv2XKSCIIB9XHZHL+ftK4/n8RiIhja79fTZDHs1rk1HcQ+ECMXL+FixdFBvb9eLnVbQQSyh/6/mqsi9zB3HU1QfA+H8/gbM9/wvxxmaw/g/NnFgyLJVJIV050FzHQ5SyES13tuLXFh3LZ8eGn2fTm+wxtTWexymZokxXlkgkglOm9FFFZv94UQa8xqOZHi533mZrq7se+s9ams5TladavVcwX7+BjOaICy05Ts2d6+Xw/aVwTAVt9jFuFdZaWOxHYTc2MFte/we9F0H/nWdPxP/uVzTqTkxcKBw8gvg5zp5kCrJ/ISRjocvXuy/jw6nZs6x1C3TM+4Hw9JmTHh4wUJqPLynwYQZFzuC1dyJO5jbdExYgDG9AlgpM2PG+X0eUqg/Vv/Ba8rju4+NbX+PI9EQQP/HvL/a+08vp+cvE1blhSGhFsbLMtg7znpo49NljvwaUjg52+3AZ8D19haMDIgKX812/hKU9xH09IV050FzHQ5WQcE8f8QH0XHt4kRh8bQm09ED52UOQRxWn29X5cbn7O+qzU/Gk0u71wpX0WSgQmu4yuYOzWvwpV++/Djd/8H1z8DeDdnpjNuVDd4MXlsQs2TWYZfj9pjn/9AyvtmwcfuU9rphwxBZaPB24g4FqNnfJCRTXTBt77Wi/84Etr5xYtiIuM61fmwJSPVfiOPE5DRutPdf7MIpgeC6Ay6RVSunKiu4+BLgfaM1nwodbUi8/95ADWwI/JInpgfNgnK3F9aMUgIn27VEk692PngZW48YvPVdPXV1h3oJAZXYbrlwFFBJrAptW2mYirZg8q5/rhn1YTlIJ9P4+sR/NOcxPgdcxqQU82w66D9+INNT2CkQVzr0ZR/uPVgHF8v1yJg5aMVQTxAZGFLcSX14YsntmTPSrNy2rbj90zzGz9yc6fxfwFjM150VCTJF9LV05UBPh4gSMZjxdk9zhB8dEfKUBCt3mz2U4RxIMdabrIUzKpzx8fLSBnYEZHRcvSHJhEVd84utGPKv51jpykOn+p/mqKlK6cqFgw0FHRMZrlrM2Bybjw1LlBtPCvc+Qo2fmbgb9vK84kzZTTlRMVDzZdEhGRozGjIyIiR2OgIyIiR2OgIyIiR2OgIyIiR2OgIyIiR2OgIyIiR2OgIyIiR2OgIyIiR2OgIyIiR2OgIyIiR2OgIyIiRyvxQPcpPn/iz/FPm7twU01ZZP7vEdgs59GHgP9TVVDaIv7m2Lvmkr84Vb5GJVV5jowXlpr/4r02rZl/WJmIio7DMzoRCA+fAv76HL77yT9rg9f3oCorbS7fCCKhgAh4jWrKcvOiMsg3BhBR8SvxQPcgNvx3GcB6sVZNsYrg/815sLrGGcEte/IVLCIYhl5ElZpSSA0NWzE2FVFjRETFqWQD3b/4fxBrjkzZdJkTvcnvxPQMTsSaBxOb5YxmQWOwNg9qTYuyaW/6eGyeZr8KCnKaKJNvd5bTT0wb67Juwyi3W3866Zo2retWw97T4qjiLPMklGlqauEZu7B4umK/vH6ssXNhou2zZTvq/Nttm4goQyUb6P6d7x/05kj/XjXF5EKXCoCH8QcE8eV/yi0gDvteg3tKZkQBnGkOoOdH5sp6N3o8g3rzoRhmO6+gNTGgjLTBdcqNWTnPVAfQdzQeyETZq+5xLRgM+3YjdGQc3RWBWIYkK/3JOn3d+vZH0ZpFhZ+uabOqL77uSGgQLfCi+5X4SzTlfrUifnxnPP2oWrT9Xaj1vI1pm+bL5Mu7UO4BLoeStHl6yvkiTyIqKGfeo3u8V92TexV/Ag/W/1q/P5e8idNei38ET5Xpn6vqTAFj/gLG5hpxpm+XmiADy3MiWIxiclpN0Ih5jDcwl5WLPQkgXr834rBPVekVHfBV6x8NMlC9YJpm2X5ByaDdBpiOVWZSkyMi8B2KH1/VoQ5Uzi0OalV1W9HzuqlTiib18jLDM2hZn+rUIgNipTu2E8IuvCADJd9iTUR5cHhnlCUyH8Jl9TGuDO4K9dHQXGu6N6ZX2ubglZLRs9EYfKOqoLAi/qNaZmrZr/mwyINFBltj2n5Nv80xC9UH0R2ctGayaZZ3ubcCwbAIsSIgBhvRoi0fQTgoEjo3QxoRFRYDXcHMIzSnPuZtBidEYEDneKzpb0l6V4pgenRsD2ZNmWmcyMhUs218MGd9BheqG65gckqNxqRYvswtsrsQIiIgouEgaj1GJuyFJaEjIioABrpcVNdqzZStpufIZjvbMGzTBFkYIvAVPKOTwfRtNJjuy8WUPY6GCpGRLWqStOeq2YPg2NtqTEi3vNaMC4SnQnDXuLRm2eGJIXGhsBXllkDHzihElL8SDXTGg+Ji8J0T4+ewoHU2+QE+T9LHobBkM+QgWmRnE9U01xrswGzB7iXtgq/Ti8t9u9X6X4O705zRmXp8agFQBF1tPqPXZrpyFZgTmxdj5fKxhHF0B+PHpw3JAk7Z0zjsCZiaNtMtL5t5R9Ez5ka1DGzywmFkVFwouAt0/oiI4lZEBfWZiIjIcdh0SUREjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZAR0REjsZARwUg3xsXfwUQOUPE32x6zdJx61vk73nqXYnGYHo3ZWmQ+3/vfKcMdDm4+fMdGNy+A5cW1ATNOCa2r8Bg77gaXzpLsv0P3seqpvcxoUYzJ99914Zg50mbt48vHWslrIblfEHr9PEifiGs/j7CE9pb2/PUPKjeDv8iqtSkZTN/Gs2m77fZX0xnW76TUn9z/plmNSkV29/XV3i1/WzS393EgCxTQ/vH+ERNN+RXLt95eQWt90iwY6DLwdof/gzbtryHD7tO4aaaFuqtR3jLAPZ37VZTls5SbH/iH3+HRw94UafGMxXxH0UPOnDSdxdemRqrhNVQsBfflgYt2Dv07evasdX0w+OPf78nMVSyrQaLfl9a4LsE/OWfqglWMkjtXdiCwNkmfH22Gv24Cu/AZ6o0/3LJ5TuJ7opRtJZcNpo9BrqcuPFw7wDWXG3HpXfF6MIp/PY8UP7MEazVZ1hiBd7+9Y/xwsX7sW/7ajUhQ+KK+2gf0P3KvRVgaKnNwN8XQIsIci9Uq0mCy/fisrYaFMyi39dnaHsJOHe2BocfUJMsPsM5MX9/x0PYrI2vxuH9IiBeDOLV64UoN7jw1CsdqBx5zfG3HRjocrXpCP6iHgj/3UFMdLXjVv151D2mypZDAbf/yXuf4/2dHhzeqCZkKDL1Ni5X7EH1osrHaDoz38dYfA9vtjPeLBXPTPRl7ZqpMs5gjCYv85WqbGoU0+Lr1bcT276lCSez/U/H/vikTNafuH/6IPffaLatEsEAc/2oipWnWkea8y+GzJs6Mzs/OX+/05MYRiNqTUHOTn7nN9XyutzPj9Xi39cDGDz7aPLWkw+u4w18G1ti83+FV3/5O/HvbQSviX/yLTcrexwNFQGMTS3+PpyEgS4P7q7zKL/qR/jqdmx7MpsmQ3FFZ7SdJwxtH6hZMpD79s0+w3/9xW381Q7bS8uUZHNSZcPjSbO5Yd9rcE/pzU5nmgPo+VG8MpGVSCviTY9nPKLC1iobF8o9wOVQQq1k8JTHtzfSZqmIYoGt7GmMTMkr1TZVOYkKzzeKys5xjGhNrLIi3I0eT3z7szb3K1LtfzrJjy8u9fkx7984uisQ23+Xb0SbLreBig7Mqm1EQiOWjGfYtxuhI/brl0GlNWhe1po9ZWKpvl+Erojjcif9fyXlf35TL1+I86PL8ffl+jb+g/xXZIM7m6YR3F+NfvMJybc8xoXqBm/y78MhGOjysXAFt9TH7MgrOtl2vngYfETNkomct28ir/5cW/Cfs9muJoJwUH1MosUfr3ir6hr1D5oZTI540X1olxoX5YdEYJp7G9Pi9yaDlkFWSEYAk5VNpdtUkyfeo+uLr08Gu5Ni2eFTp/GTzjYMi3n1ICfMX8DYXCPOmOZ3+Z5DC0YxabpqT77/6aQ+PkPq8yPK64zlc6uMzE1/tvufsD/ZWvLvV2OXlRXi/KZfPt/zo8n59wX8r7emsOrZz7Hvx/b1Qr7lFsGw5SLBaRjochbCJdlkuMWHctkx5KfZ9HYsREaXz/YNepPGX+032vKXyXwYQYgr7BqjAhNDTT8uq2KXe6v64YkKKdiIluCkyLT0wOpxp7rOt9JutqMfPSPWoIb5UGxbcWVwi6ypINIcX3q7UNssMpIJo+k1gukxuyCQO5kVznYivo8J2VBeCvH9zoXU/hi9GwfFhYiS7/nNYPnCnJ88fl+Rq9j7mw0IyPt4WhPkHxA070C+5fcYBrpcvfsyPpRNhr1DqHvGB5yvx4TsGJKRAmR0eW1fuX4Nb0X+FHtzuNrMn7iiVs1K8UFdgZe5xdW1qOhEhYSGg6j1GJmWF9nU9VqPUM8gzjSPojVtRTWP0Jz6WBApji9TsaZZvRkzlpEWiNEEqjWNiguCqoL2vsv9+3XV7EFlQna9WL7nN/3yeZ+fXH9ff/ZtPCr+sQTI63/AR1DryrfcjvmWgAMx0OVkHBPH/EB9Fx7eJEYfG0Kt7Bhy7KDIs5ZDYbY/8aurQA6PFOhUc9rYheyvdNUN8J7Xk1QcZeXwiH/CUyG4a1xas9PwxJAIRFtRnmlFNn0cVX1btUyuqk9kA3P9OGp0gKiu1Zopzd2qZ2XzZkUHfDndh0mQ7vjSUk1r5orYnJEqWmZUiOY18V3K+2YFk+/3W/Y0DsuM1re4A4km3/Ob9fK5nZ+cf18bH8KJncAbvzSefROZYf9VvL9zo76ufMstCt9aUIwY6HKgPbMGH2q74h1A3E8OYA38mFyGB8YLs329C3LWjxSYaFfeIoD4U15523HhqXPiKjmY0JkklnXJZsRR9Iy59R6dMjCNjIpAlNBBIbEzirG87GGpdT45qB5ylg/HiqDct1t1TpHNYSL4mZbXOh5k+xyepcejHIyKOd3xpSP319Rslmz56hf1Thax+ZIEBhvavTHTurXjtwmmucn/+5U9Si1Nh25xIaLK8j+/6ZfP//yk+n2JwKM9KC6Gl2RvyN9hr3brYirW/b+uXT37pk2fRsembfi6Pd6hJd/yGO1+tRcN4oLD0aJ0Twq9ORldeXJBjeVu5pgn+uATb0Tn1TgVQPiNaNPmpujpsBrXTEePb/ZEm4aW70zPDzVFHzw2rcYoGe03kHCeCvX7Wlrz0dNPLO//qbuFGd09KfdHChJV9Ykr44Lf37nH2XWW0TpQZNcZh+6Wwv2+ltJd/atGy2yFjHbqM1GOZBdw+cxSlp0tKCntOa8RNaIk/qWQpSafJdMeStc04szd+HuXRUv+nzc1p8pHXQrW9Lsc5P5PovYe+U4Z6IiIyNHYdElERI7GQEdERI7GQEdERI7GQEdERI7GQEdERI7GQEdERI7GQEdERA4G/H8u17jZnE71SAAAAABJRU5ErkJggg=="
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image.png](attachment:image.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_counts = exonCounts / (exonLengths / 1000)\n",
    "log_norm_counts = np.log2(norm_counts + 1)\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "axs[0].scatter(\n",
    "    np.mean(log_norm_counts, axis=0),\n",
    "    np.var(log_norm_counts, axis=0),\n",
    "    s=10,\n",
    ")\n",
    "axs[0].plot(\n",
    "    np.arange(0, np.max(np.mean(log_norm_counts, axis=0)), 1),\n",
    "    np.arange(0, np.max(np.mean(log_norm_counts, axis=0)), 1),\n",
    "    color=\"r\",\n",
    "    linewidth=2,\n",
    "    label=\"y = x\",\n",
    ")\n",
    "\n",
    "axs[0].set_xscale(\"log\")\n",
    "axs[0].set_yscale(\"log\")\n",
    "axs[0].set_xlabel(\"Mean Expression\")\n",
    "axs[0].set_ylabel(\"Expression Variance\")\n",
    "axs[0].set_title(\"Gene variance vs mean expression\")\n",
    "axs[0].tick_params(which=\"minor\", length=1, color=\"k\")\n",
    "axs[0].tick_params(which=\"major\", length=5, color=\"k\")\n",
    "axs[0].legend()\n",
    "\n",
    "\n",
    "# add plot\n",
    "axs[1].scatter(\n",
    "    np.mean(log_norm_counts, axis=0),\n",
    "    np.var(log_norm_counts, axis=0) / np.mean(log_norm_counts, axis=0),\n",
    "    s=10,\n",
    ")\n",
    "\n",
    "axs[1].plot(\n",
    "    np.arange(0, np.max(np.mean(log_norm_counts, axis=0)), 1),\n",
    "    np.ones(np.arange(0, np.max(np.mean(log_norm_counts, axis=0)), 1).size),\n",
    "    color=\"r\",\n",
    "    linewidth=2,\n",
    "    label=\"Model: F = 1\",\n",
    ")\n",
    "\n",
    "axs[1].set_xlabel(\"Mean Expression\")\n",
    "axs[1].set_ylabel(\"Fano Factor\")\n",
    "axs[1].set_title(\"Normalized Counts: Fano Factor vs Mean Expression\")\n",
    "axs[1].set_ylim(0.01, 20)\n",
    "axs[1].set_xscale(\"log\")\n",
    "axs[1].set_yscale(\"log\")\n",
    "axs[1].tick_params(which=\"minor\", length=2, color=\"k\")\n",
    "axs[1].tick_params(which=\"major\", length=5, color=\"k\")\n",
    "axs[1].legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "depth_median = np.median(depth)\n",
    "norm_factor = depth_median * (1 / depth)\n",
    "\n",
    "norm_counts = norm_factor[:, np.newaxis] * exonCounts\n",
    "# norm_counts = exonCounts / (exonLengths / 1000)\n",
    "log_norm_counts = np.log2(norm_counts + 1)\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "axs[0].scatter(\n",
    "    np.mean(log_norm_counts, axis=0),\n",
    "    np.var(log_norm_counts, axis=0),\n",
    "    s=10,\n",
    ")\n",
    "axs[0].plot(\n",
    "    np.arange(0, np.max(np.mean(log_norm_counts, axis=0)), 1),\n",
    "    np.arange(0, np.max(np.mean(log_norm_counts, axis=0)), 1),\n",
    "    color=\"r\",\n",
    "    linewidth=2,\n",
    "    label=\"y = x\",\n",
    ")\n",
    "\n",
    "axs[0].set_xscale(\"log\")\n",
    "axs[0].set_yscale(\"log\")\n",
    "axs[0].set_xlabel(\"Mean Expression\")\n",
    "axs[0].set_ylabel(\"Expression Variance\")\n",
    "axs[0].set_title(\"Gene variance vs mean expression\")\n",
    "axs[0].tick_params(which=\"minor\", length=1, color=\"k\")\n",
    "axs[0].tick_params(which=\"major\", length=5, color=\"k\")\n",
    "axs[0].legend()\n",
    "\n",
    "\n",
    "# add plot\n",
    "axs[1].scatter(\n",
    "    np.mean(log_norm_counts, axis=0),\n",
    "    np.var(log_norm_counts, axis=0) / np.mean(log_norm_counts, axis=0),\n",
    "    s=10,\n",
    ")\n",
    "\n",
    "axs[1].plot(\n",
    "    np.arange(0, np.max(np.mean(log_norm_counts, axis=0)), 1),\n",
    "    np.ones(np.arange(0, np.max(np.mean(log_norm_counts, axis=0)), 1).size),\n",
    "    color=\"r\",\n",
    "    linewidth=2,\n",
    "    label=\"Model: F = 1\",\n",
    ")\n",
    "\n",
    "axs[1].set_xlabel(\"Mean Expression\")\n",
    "axs[1].set_ylabel(\"Fano Factor\")\n",
    "axs[1].set_title(\"Normalized Counts: Fano Factor vs Mean Expression\")\n",
    "axs[1].set_ylim(0.01, 20)\n",
    "axs[1].set_xscale(\"log\")\n",
    "axs[1].set_yscale(\"log\")\n",
    "axs[1].tick_params(which=\"minor\", length=2, color=\"k\")\n",
    "axs[1].tick_params(which=\"major\", length=5, color=\"k\")\n",
    "axs[1].legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA vizualization for different transformations and normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform PCA three times for each normalization: on the resulting matrix as is,\n",
    "# after np.log2(X+1) transform, and after np.sqrt(X) transform\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "\n",
    "Fano = np.var(log_CTF_exon_counts, axis=0) / np.mean(\n",
    "    log_CTF_exon_counts, axis=0\n",
    ")\n",
    "Selected_genes = Fano > 5\n",
    "print(np.sum(Selected_genes))\n",
    "# CTF\n",
    "sqrt_CTF = np.sqrt(CTF_exon_counts)\n",
    "\n",
    "# perform PCA\n",
    "pca = PCA(n_components=50)\n",
    "pca_CTF = pca.fit_transform(CTF_exon_counts[:, genesForRRR])\n",
    "pca_log_CTF = pca.fit_transform(log_CTF_exon_counts[:, genesForRRR])\n",
    "pca_sqrt_CTF = pca.fit_transform(sqrt_CTF[:, genesForRRR])\n",
    "\n",
    "# Gene lenght norm\n",
    "Fano = np.var(log_norm_counts, axis=0) / np.mean(log_norm_counts, axis=0)\n",
    "Selected_genes = Fano > 6\n",
    "print(np.sum(Selected_genes))\n",
    "\n",
    "sqrt_norm = np.sqrt(norm_counts)\n",
    "\n",
    "# perform PCA\n",
    "pca = PCA(n_components=50)\n",
    "pca_norm_counts = pca.fit_transform(norm_counts[:, Selected_genes])\n",
    "pca_log2__norm_counts = pca.fit_transform(log_norm_counts[:, Selected_genes])\n",
    "pca_sqrt_norm_counts = pca.fit_transform(sqrt_norm[:, Selected_genes])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(2, 3, figsize=(9, 6))\n",
    "axs = axs.flatten()\n",
    "\n",
    "# Plot CTF\n",
    "\n",
    "axs[0].scatter(pca_CTF[:, 0], pca_CTF[:, 1], s=3, c=cluster_colors, alpha=0.8)\n",
    "axs[0].set_title(\"PCA\")\n",
    "\n",
    "axs[1].scatter(\n",
    "    pca_log_CTF[:, 0], pca_log_CTF[:, 1], s=3, c=cluster_colors, alpha=0.8\n",
    ")\n",
    "axs[1].set_title(\"PCA after $log_2(X+1)$\")\n",
    "\n",
    "axs[2].scatter(\n",
    "    pca_sqrt_CTF[:, 0], pca_sqrt_CTF[:, 1], s=3, c=cluster_colors, alpha=0.8\n",
    ")\n",
    "axs[2].set_title(\"PCA after sqrt\")\n",
    "\n",
    "axs[3].scatter(\n",
    "    pca_norm_counts[:, 0],\n",
    "    pca_norm_counts[:, 1],\n",
    "    s=3,\n",
    "    c=cluster_colors,\n",
    "    alpha=0.8,\n",
    ")\n",
    "axs[3].set_title(\"PCA\")\n",
    "\n",
    "axs[4].scatter(\n",
    "    pca_log2__norm_counts[:, 0],\n",
    "    pca_log2__norm_counts[:, 1],\n",
    "    s=3,\n",
    "    c=cluster_colors,\n",
    "    alpha=0.8,\n",
    ")\n",
    "axs[4].set_title(\"PCA after $log_2(X+1)$\")\n",
    "axs[4].set_xlim(-50, 100)\n",
    "axs[4].set_ylim(-75, 75)\n",
    "\n",
    "axs[5].scatter(\n",
    "    pca_sqrt_norm_counts[:, 0],\n",
    "    pca_sqrt_norm_counts[:, 1],\n",
    "    s=3,\n",
    "    c=cluster_colors,\n",
    "    alpha=0.8,\n",
    ")\n",
    "axs[5].set_title(\"PCA after sqrt\")\n",
    "axs[5].set_xlim(-300, 300)\n",
    "axs[5].set_ylim(-200, 200)\n",
    "\n",
    "\n",
    "fig.supxlabel(\"$1^{st}$ PC\")\n",
    "fig.supylabel(\"$2^{nd}$ PC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tsne = TSNE(perplexity=30)\n",
    "tsne_counts = tsne.fit(pca_sqrt_norm_counts)\n",
    "fig, axs = plt.subplots()\n",
    "axs.scatter(\n",
    "    tsne_counts[:, 0], tsne_counts[:, 1], s=20, c=cluster_colors, alpha=0.8\n",
    ")\n",
    "axs.set_title(\"TSNE\")\n",
    "axs.set_axis_off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "?TSNE"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
